proposals/NNNN-explicit-padding.md

```diff
@@ -0,0 +1,175 @@
+can generally describe the appropriate type representations for these using
+explicit padding, but some care needs to be taken:
+
+- Arrays in CBuffers may have padding in between members, but crucially they do
+  not have padding after the last member. This needs special handling to
+  represent.
+
+- There are two HLSL constructs that may introduce arbitrary padding to a
+  struct. In a cbuffer, the `packoffset` attribute specifies the offset of a
+  member, and outside of cbuffers, the `vk::offset` vulkan attribute may do the
+  same.
+
+- Simply padding structures with `i8` as is typical with ABI-related padding
+  makes it difficult to recover which struct elements are padding vs which are
+  subobjects. This matters in some backends, and is specifically important for
+  SPIR-V where we need to map a logical indices into the struct into physical
+  offsets.
+
+## Proposed solution
+
+Introduce an explicit padding type and use this type for the padding in the
+various constructs that need it.
+
+The padding type will be a target type with a parameter for the size in bytes.
+For DirectX, this would look like `target("dx.Layout", 4)` for 4 bytes of
+padding. For SPIR-V, `target("spirv.Padding", 8)` for 8 bytes of padding. We
+may attempt to come up with a first-class type in LLVM for these purposes in
+the future.
+
+### Structs with annotations
+
+Generally speaking HLSL structs are equivalent to packed structs in C++. We can
+simply add padding between members as appropriate in order to satisfy the rules
+specified by `packoffset`, `vk::offset`, or otherwise by HLSL semantics.
+
+There is one complicating factor here. Both `dxc` and `fxc` allow `packoffset`
+to be used to layout a struct in an order that does not match the
+lexicographical order:
+
+```hlsl
+cbuffer cb0 {
+    int x : packoffset(c0.y);
+    float y : packoffset(c0.x);
+}
+```
+
+To support this, we need to create the underlying LLVM type in the order that
+matches the packoffsets rather than the order as written, so we would end up
+with `{ float, i32 }` here, losing the lexicographical order. This is probably
+okay since we need to create artificial types for cbuffers anyway (such as when
+we filter out resource types that are declared within the cbuffer), but may not
+make for a particularly good debugging experience.
+
+### Arrays in a cbuffer
+
+CBuffers in HLSL have very specific layout rules. Each "object" starts at a
+16-byte boundary, which is mostly explainable as a 16-byte alignment
+requirement, but applies to array elements rather than the array itself in a
+way that doesn't match the general language. We can emulate this with an array
+of objects that consist of a struct containing the element type and padding to
+16 bytes, followed by a single instance of the element type itself. See
+[CBuffer Padded arrays at the HLSL-level] for details.
+
+[CBuffer Padded arrays at the HLSL-level]: #cbuffer-padded-arrays-at-the-hlsl-level
+
```

Comment:
Add section "### Struct in a cbuffer" (here or before the array section) describing that structs in cbuffer must also start on a 16-byte boundary and we need to insert padding before if as needed.

Reply:
I found a separate section for this more confusing so I instead reworked this a bit to talk about scalars, vectors, structs, and arrays explicitly.